src/dataflow/src/sink/tail.rs

```diff
@@ -66,6 +70,29 @@ pub fn tail<G>(
             // though it is slower because it will produce deterministic results since the
             // cursor will always produce rows in the same order.
             results.sort_by_key(|(time, _)| *time);
+
+            if connector.emit_timestamps && !batches.is_empty() {
+                // The user has request timestamp done messages. We've already sorted the
+                // results by timestamp so we can duplicate the last row's timestamp which is
+                // somewhat easier than using a batch.upper() antichain. This timestamp won't
+                // be reused any other batch because they are disjoint and strictly increasing.
+                let (time, _) = results.last().unwrap().clone();
+                packer.push(Datum::Decimal(Significand::new(i128::from(time))));
+                packer.push(Datum::True);
```

Comment:
I don't know anything about the downstream protocol but it seems like the first two fields of this timestamp message / progress message could be totally different types than the rest of the sink -- is that just fine / handled nicely by COPY? asking mostly for my own edification

Reply:
ignore me! this is the same confusion as the last comment